Modify action to build and consume container

weierophinney opened the pull request:
To speed things up further, we can build and release the actual container on release publication, and then have the action itself consume the image instead of the Dockerfile.

Additionally, we can force the release process to use the latest set of changes by defining a local GHA that uses the Dockerfile instead.
I've proven this works at https://github.com/weierophinney/test-local-action/, where I observed that the build happens relative to the checkout directory, and not to where the action is defined.

This will require defining two secrets in this repository:

- CONTAINER_USERNAME (username on ghcr.io)
- CONTAINER_PAT (personal authentication token with container write permissions)

Once setup, this should automatically build and push tags for the docker image, and the action will use the image to do its work.

weierophinney committed:
feat: modify action to build and consume container

To speed things up further, we can build and release the actual container on release publication, and then have the action itself consume the image instead of the Dockerfile.

Additionally, we can force the release process to use the latest set of changes by defining a local GHA that uses the Dockerfile instead.
I've proven this works at https://github.com/weierophinney/test-local-action/, where I observed that the build happens relative to the checkout directory, and not to where the action is defined.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
I've added the secrets at this point. After first build, we'll need to make the container public, and then associate it with this repository; I'll go ahead and do that when released.

weierophinney commented on .github/workflows/build-and-push-containers.yml:
```diff
@@ -0,0 +1,47 @@
+          MAJOR="${PREFIX}:$(echo ${TAG} | cut -d. -f1)"
+          MINOR="${MAJOR}.$(echo ${TAG} | cut -d. -f2)"
+          PATCH="${PREFIX}:${TAG}"
+          echo "::set-output name=tags::${MAJOR}%0A${MINOR}%0A${PATCH}"
```
GHA replaces `%0A` with `\n`, making these line-delimited tags. This creates a list that looks like:

```text
ghcr.io/laminas/automatic-releases:1
ghcr.io/laminas/automatic-releases:1.11
ghcr.io/laminas/automatic-releases:1.11.0
```

which is then passed as the variable "tags" to the next step.

glensc replied:
This is probably not the best idea to rely on undocumented behavior. Seems like an implementation detail that may behave differently in the future.

However, they have documented to pass as json, if you need something like this:

```yaml
name: build
on: push
jobs:
  job1:
    runs-on: ubuntu-latest
    outputs:
      matrix: ${{ steps.set-matrix.outputs.matrix }}
    steps:
    - id: set-matrix
      run: echo "::set-output name=matrix::{\"include\":[{\"project\":\"foo\",\"config\":\"Debug\"},{\"project\":\"bar\",\"config\":\"Release\"}]}"
  job2:
    needs: job1
    runs-on: ubuntu-latest
    strategy:
      matrix: ${{fromJSON(needs.job1.outputs.matrix)}}
    steps:
    - run: build
```

- https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#example-returning-a-json-object

weierophinney replied:
@glensc This is actually the approach recommended by the GHA team, and I discovered it here: https://github.community/t/set-output-truncates-multiline-strings/16852/3 (they are indicating that they transform ANSI escape sequences emitted in outputs).

A later note in that thread indicates that it's undocumented, but that they will be updating the docs to reflect this.

weierophinney replied:
I've done some thinking on this, and decided to go with the JSON approach; I'm now doing:

```yaml
with:
  tags: ${{ join(fromJSON(steps.tags.outputs.tags), "\n") }}
```

(as they "tags" value is expected to be multiline).

This should address any concerns about ANSI escape sequences, and ensure we are forwards-compatible with any changes GHA makes.

weierophinney replied:
Update: using "," as the separator, as CSV is supported for that value, and the `"\n"` value was causing issues with the GHA expressions.

weierophinney commented on .github/actions/automatic-releases/action.yml:
```diff
@@ -0,0 +1,20 @@
+       * `laminas:automatic-releases:switch-default-branch-to-next-minor`
+    required: true
+
+runs:
+  using: 'docker'
+  image: '../../../Dockerfile'
+  args:
+    - ${{ inputs.command-name }}
```
What I discovered is that while the image is in a directory relative to this one, the actual build happens relative to `$GITHUB_WORKSPACE`, so everything works correctly. By doing this, we get the benefits of whatever changes we've made for this release... when creating the release.

weierophinney commented on action.yml:
```diff
@@ -15,6 +15,6 @@ inputs:
 runs:
   using: 'docker'
-  image: 'Dockerfile'
+  image: 'docker://ghcr.io/laminas/automatic-releases:1'
```
This is the magic: the action now uses an image on ghcr.io, which means only a pull operation is required, and no build operations!

weierophinney commented on .github/workflows/automatic-release.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
 
       - name: "Release"
-        uses: "./"
+        uses: "./.github/actions/automatic-releases/"
```
And this part ensures we use that local action, which builds from the Dockerfile directly instead of using the image.

weierophinney committed:
refactor: use JSON outputs instead of multiline

To address concerns with using ANSI escape sequences to create multiline outputs, this patch does the following:

- Emits a JSON array instead.
- Creates a single job, combining steps from each previous job; this prevents the need to use `needs`, making it possible to test the solution locally.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: use CSV instead of multi-line list for tags input

GHA does not like newlines in strings used as part of the expression syntax.
Since docker/build-push-action allows CSV, comma separated will accomplish the same goals.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

bendavies commented on .github/workflows/build-and-push-containers.yml:
```diff
@@ -0,0 +1,39 @@
+  release-container:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Compile tag list
```
you could use https://github.com/crazy-max/ghaction-docker-meta here

weierophinney replied:
True - but this is fast and doesn't require any additional overhead as it's using tooling already in the environment.

Thanks for the link, though - I'd not seen that one!

Ocramius commented on .github/workflows/build-and-push-containers.yml:
```diff
@@ -0,0 +1,39 @@
+          PATCH="${PREFIX}:${TAG}"
+          echo "::set-output name=tags::[\"${MAJOR}\",\"${MINOR}\",\"${PATCH}\"]"
+      - name: Checkout
+        uses: actions/checkout@v2
```
Not super-important, but some spacing (blank lines around blocks) would go a long way with improving readability

weierophinney replied:
Done!

Ocramius commented on .github/workflows/build-and-push-containers.yml:
```diff
@@ -0,0 +1,39 @@
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ secrets.CONTAINER_USERNAME }}
```
These secrets should be prefixed, to make it clear they are local to our build

weierophinney replied:
Done; used `AUTOMATIC_RELEASES_` as the prefix.

Ocramius commented on .github/workflows/build-and-push-containers.yml:
```diff
@@ -0,0 +1,39 @@
+          file: ./Dockerfile
+          platforms: linux/amd64
+          push: true
+          tags: ${{ join(fromJSON(steps.tags.outputs.tags), ",") }}
```
So we push each image as 3 tags? Ok for me, hard to read/follow if not versed in github actions (requires comments around it)

weierophinney replied:
Yes - actual release version, minor release version, major release version.

Rationale is that if we change something in a new minor release that causes issues for somebody, they can pin to the previous minor. Alternately, they can pin to a specific "known good" release. (There are existing actions for creating tags that do the same; I chose to do it using standard *nix tooling, as it introduces zero additional overhead.)

Ocramius reviewed (approved):
Besides minor improvement suggestions, LGTM :+1:

weierophinney committed:
qa: prefix secrets to make clear they are specific to this repo

Prefixed with `AUTOMATIC_RELEASES_`

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: whitespace between job steps in container build workflow

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

glensc commented on .github/workflows/build-and-push-containers.yml:
```diff
@@ -0,0 +1,39 @@
+      - name: Compile tag list
+        id: tags
+        run: |
+          TAG=${GITHUB_REF/refs\/tags\//}
```
```suggestion
          TAG=${GITHUB_REF#refs/tags/}
```

I'm sure I've already suggested this elsewhere.

weierophinney replied:
Done.

weierophinney committed:
qa: simplify bash pattern matching when matching tag name

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
